### PR TITLE
2023 workflows up to tracking reco

### PR DESCRIPTION
**Pull request by ebrondol**
With this PR, the tracking reconstruction has been introduced in the 2023 scenario with both the tilted (2023tilted workflow) and the flat (2023GReco workflow) tracker geometry. 
In order to fix also the Lorentz Angle message for the tilted, this PR should be integrated together with cms-data/SLHCUpgradeSimulations-Geometry#2.

@boudoul @delaere @VinInn @rovere

### CODE REVIEW

**Comment by slava77** [comment]
@ebrondol @boudoul 
Can this PR just wait for https://github.com/cms-data/SLHCUpgradeSimulations-Geometry/pull/2 to be in a CMSSW IB?

It's more difficult to get things in order if both cms-dist and cms-sw have to be updated at the same time.

**Comment by ebrondol** [comment]
@slava77 
I have nothing against it.

**Comment by slava77** [comment]
@ebrondol 
If I understood your comment correctly, we can wait for the cms-data/SLHCUpgradeSimulations-Geometry#2 to be in a CMSSW IB and then test this PR.

**Comment by ebrondol** [comment]
@slava77 
In principle, yes. Do we know how long it will take for the other PR to be merged?

File: Configuration/Geometry/python/GeometryExtended2023RecoReco_cff.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import FWCore.ParameterSet.Config as cms
```
Comment (ianna) [inline Configuration/Geometry/python/GeometryExtended2023RecoReco_cff.py]: @ebrondol - Is this file needed?

Comment (boudoul) [inline Configuration/Geometry/python/GeometryExtended2023RecoReco_cff.py]: Hi @ebrondol , @ianna , this file (GeometryExtended2023RecoReco_cff.py) is indeed not needed, probably a remant of the branch when we were preparing the PR and debating the geometry name to be used - but indeed not needed anymore

**Comment by boudoul** [comment]
thanks to @degano , cms-data/SLHCUpgradeSimulations-Geometry#2 is now merged, so it should be included in the next IB

**Comment by boudoul** [comment]
just adding @kpedro88 , @atricomi, @emiglior  so that thy can watch as well

**Comment by ianna** [comment]
@ebrondol  and @boudoul - these configuration files are identical:
```
Configuration/Geometry/python/GeometryExtended2023LReco_cff.py
Configuration/Geometry/python/GeometryExtended2023GReco_cff.py
Configuration/Geometry/python/GeometryExtended2023sim_cff.py
```
so are these:
```
Configuration/Geometry/python/GeometryExtended2023LRecoReco_cff.py
Configuration/Geometry/python/GeometryExtended2023GRecoReco_cff.py
Configuration/Geometry/python/GeometryExtended2023simReco_cff.py
```
IMHO, it is better to use one geometry configuration even if you want to use it with different labels. If they diverge, then it would be fine to introduce the new ones. Otherwise they may diverge unintentionally.

File: Configuration/StandardSequences/python/GeometryConf.py [hunk 2]
```diff
@@ -33,6 +33,6 @@
     'HCal'        : 'Configuration.Geometry.GeometrySimHCAL_cff,Configuration.Geometry.GeometryRecoHCAL_cff',
     'Extended2023sim'   : 'Extended2023sim,Extended2023simReco',
     'Extended2023tilted'   : 'Extended2023tilted,Extended2023tiltedReco',
-    'Extended2023LReco'   : 'Extended2023LReco,Extended2023LRecoReco'
-    
+    'Extended2023LReco'   : 'Extended2023LReco,Extended2023LRecoReco',
+    'Extended2023GReco'   : 'Extended2023GReco,Extended2023GRecoReco'
```
Comment (ianna) [inline Configuration/StandardSequences/python/GeometryConf.py]: @ebrondol and @boudoul - by introducing these labels and pointing them to existing scenario you'd avoid file multiplication:
```
'Extended2023LReco'   : 'Extended2023sim,Extended2023simReco',
'Extended2023GReco'   : 'Extended2023sim,Extended2023simReco',
```

**Comment by ebrondol** [comment]
@ianna @boudoul @kpedro88 
Please, let me know if the last changes are fine for you.

**Comment by ianna** [comment]
@ebrondol - Thanks, it looks fine.

**Comment by ianna** [comment]
+1

**Comment by civanch** [comment]
+1

**Comment by boudoul** [comment]
Hi @fabozzi  @srimanob   @hengne , could you please check the PdmV part of this PR (runthematrix) and eventually sign it ? This PR  is an important milestone for the phase2  - Thanks

**Comment by ebrondol** [comment]
About the demonstration of this PR, you shouId get the PR in the last release and running the WF#11200 and #10600. Both produce a step3.root, which contains pixelTracks (which can be seen via Fireworks). In attachment an example.
pixeltracks_wf10600
pixeltracks_wf11200

**Comment by slava77** [comment]
@ebrondol @boudoul @kpedro88 
Do we have minBias gensim that can go with a setup with pileup?
All matrix workflows that I've see for 2023 with ```-w upgrade``` were without PU

**Comment by boudoul** [comment]
Good point - we do actually- I can setup the WFs by tomorrow

File: RecoTracker/TkDetLayers/src/Phase2OTtiltedBarrelLayer.cc [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+      it!=theComps.end();it++){  
+    theBasicComps.insert(theBasicComps.end(),	
+			 (**it).basicComponents().begin(),
+			 (**it).basicComponents().end());
+  }
+
+  BarrelDetLayer::initialize();
+  theCylinder = cylinder( theComps );
+
+  //--------- DEBUG INFO --------------
```
Comment (slava77) [inline RecoTracker/TkDetLayers/src/Phase2OTtiltedBarrelLayer.cc]: since LogTrace and LogDebug are compiled only if EDM_ML_DEBUG is defined,
it makes sense to put all code that does only LogDebug or LogTrace inside some loops altogether inside ``` #ifdef EDM_ML_DEBUG ```

File: RecoTracker/TkDetLayers/src/Phase2OTtiltedBarrelLayer.cc [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+  }
+  for(auto ring : thePositiveRingsComps){
+    ring->groupedCompatibleDetsV(tsos, prop, est, closestResultPos);
+  }
+
+  result.assign(closestResultRods.begin(),closestResultRods.end());
+  result.insert(result.end(),closestResultPos.begin(),closestResultPos.end());
+  result.insert(result.end(),closestResultNeg.begin(),closestResultNeg.end());
+
+  LogDebug("TkDetLayers") << "==== output di Phase2OTtiltedBarrelLayer =====" ;
```
Comment (slava77) [inline RecoTracker/TkDetLayers/src/Phase2OTtiltedBarrelLayer.cc]: everything below is debugs not compiled anyways except for a bunch of empty loops
put this inside ```#ifdef EDM_ML_DEBUG ```

File: RecoTracker/TkDetLayers/src/Phase2OTtiltedBarrelLayer.h [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+#include "Phase2OTBarrelRod.h"
+#include "Phase2OTEndcapRing.h"
+#include "SubLayerCrossings.h"
+
+
+/** A concrete implementation for Phase2OTtiltedBarrel layer 
+ *  built out of BarrelPhase2OTBarrelRod
+ */
+
+#pragma GCC visibility push(hidden)
```
Comment (slava77) [inline RecoTracker/TkDetLayers/src/Phase2OTtiltedBarrelLayer.h]: wouldn't  ``` namespace { ``` give the same effect of internal linkage?

Comment (VinInn) [inline RecoTracker/TkDetLayers/src/Phase2OTtiltedBarrelLayer.h]: "namespace {"  defines it in the context of the file (equivalent to inline)
hidden visibility in the context of shared library

File: RecoTracker/TkDetLayers/src/Phase2OTtiltedBarrelLayer.h [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+			       const MeasurementEstimator& est,
+			       std::vector<DetGroup> & result) const final;
+    
+  //virtual SubDetector subDetector() const { return GeomDetEnumerators::subDetGeom[GeomDetEnumerators::P2OTB];}
+
+ private:
+  //std::vector<const GeometricSearchDet*> theComps;
+  std::vector<const GeometricSearchDet*> theNegativeRingsComps;
+  std::vector<const GeometricSearchDet*> thePositiveRingsComps;
+  std::vector<const GeomDet*> theBasicComps;
```
Comment (slava77) [inline RecoTracker/TkDetLayers/src/Phase2OTtiltedBarrelLayer.h]: there are also theBasicComps in the base class.
Is it needed here as well?

File: RecoTracker/TkDetLayers/src/TBLayer.h [hunk 7]
```diff
@@ -25,14 +25,14 @@ class TBLayer: public BarrelDetLayer {
 
   // GeometricSearchDet interface
   
-  virtual const std::vector<const GeomDet*>& basicComponents() const final {return theBasicComps;}
+  virtual const std::vector<const GeomDet*>& basicComponents() const {return theBasicComps;}
```
Comment (slava77) [inline RecoTracker/TkDetLayers/src/TBLayer.h]: is removal of finalization justified?
I see there is a formal overload (in an apparently derived from this) Phase2OTtiltedBarrelLayer, but the overload uses the same ``` return theBasicComps ```, except that those theBasicComps are also a data member in the derived

Comment (VinInn) [inline RecoTracker/TkDetLayers/src/TBLayer.h]: we know.
Need to be refactorized.
next iteration.

Comment (ebrondol) [inline RecoTracker/TkDetLayers/src/TBLayer.h]: This is actually something I can already do now. I will have a look and modify.

Comment (slava77) [inline RecoTracker/TkDetLayers/src/TBLayer.h]: if it's in plans, maybe easier to leave as is. 
I don't think the CPU cost is significant from this to the run2 tracking.
Vincenzo, any good guess by how much it would slow down (having in mind all de-finalization here)?

Comment (ebrondol) [inline RecoTracker/TkDetLayers/src/TBLayer.h]: This specific line was not in plan, I think. I just forgot about cleaning up this lines.
The rest of the methods that are not final anymore, they should stay like that - for now, as Vincenzo said.

Comment (slava77) [inline RecoTracker/TkDetLayers/src/TBLayer.h]: this one can't be final because there is an overload in the derived Phase2OTtiltedBarrelLayer.
so, unless that overload can be removed ...

Comment (ebrondol) [inline RecoTracker/TkDetLayers/src/TBLayer.h]: Sorry for the misunderstanding. This discussion can be related with your other comment 
"there are also theBasicComps in the base class.
Is it needed here as well?"
and the answer is yes, I am going to remove the methods from Phase2OTtiltedBarrelLayer and keeping the "final" in TBLayer.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
trying in CMSSW_8_1_X_2016-05-18-2300

11024 is broken in two different ways
baseline:
```
   [2] Calling event method for module MixingModule/'mix'
Exception Message:
Cannot initialize HcalDetId from 3
```

with this PR:
```
   [2] Calling event method for module MixingModule/'mix'
Exception Message:
A std::exception was thrown.
vector::reserve
```

I see that it fails in IB as well, curiously enough, with the same message as in my test with this PR.
So, likely not related to parts of this PR.
This is 2023LReco workflow, not 2023GReco of this PR. Still, something to keep in mind.

**Comment by kpedro88** [comment]
@slava77 - the vector::reserve exception is my fault, bug fix PR is here: #14572

Not sure about "cannot initialize HcalDetId from 3", I did not see that when running upgrade WF 11000.0 in CMSSW_8_1_X_2016-05-18-1100.

**Comment by slava77** [comment]
On 5/19/16 10:31 AM, Kevin Pedro wrote:
> @slava77 <https://github.com/slava77> - the vector::reserve exception is
> my fault, bug fix PR is here: #14572
> <https://github.com/cms-sw/cmssw/pull/14572>
> 
> Not sure about "cannot initialize HcalDetId from 3", I did not see that
> when running upgrade WF 11000.0 in CMSSW_8_1_X_2016-05-18-1100.

Since it's a "weird memory issue", I guess it can flip to different ways
of crashing at will

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/14523	9fb639c
- code changes are in line with the description: tracking reco for phase2 still based on customization function. 2023LReco now goes up to siPixelRecHits; non-functional 2023Reco is replaced by 2023GReco and goes up to tracking.
- jenkins tests pass and comparisons with baseline show no differences (this uses run1, run2 and 2017 workflows)
- local tests using 11024 (2023LReco), as well as  11200 (4mu gun changed from 2023Reco to 2023GReco .. flat tracker layout) and 10600 (4mu 2023tilted reco up to tracking) shows that these workflows run [CMSSW_8_1_X_2016-05-18-2300 +#14572 as a baseline]:
    - 11024: siPixelRecHits and siPixelClusters are added at run time, but only siPixelClusters are saved. Supposedly, the hits will remain transient (current CPE is not dependent on track parameters)
    - 11200: didn't run in the baseline, but runs now; the output is RECOSIM
    - 10600: runs in both, now it contains tracking reco modules. The output changes from FEVTDEBUGHLT to RECOSIM; step3 changes from ``` RAW2DIGI,L1Reco,RECO:localreco  ``` to ``` RAW2DIGI,L1Reco,RECO:reconstruction_trackingOnly```

**Comment by mdhildreth** [comment]
+1

**Comment by kpedro88** [comment]
@hengne please sign if it's acceptable...

**Comment by smuzaffar** [comment]
@ebrondol , as mentioned in #14921, this PR has caused the failure of one unit test. Can you please provide a fix for that

**Comment by ebrondol** [comment]
Hi @smuzaffar, as far as I understand from the error, it is just a matter to change `GeometryExtended2023LRecoReco_cff` with `GeometryExtended2023simReco_cff`. We introduce that commit after @ianna comment, which was suggesting to use one geometry configuration for now, since all the files were the same.

**Comment by ianna** [comment]
@ebrondol and @smuzaffar - I think, the fix is already merged.

**Comment by smuzaffar** [comment]
@ianna , do you know the PR which has the fix?

**Comment by smuzaffar** [comment]
These files
https://github.com/cms-sw/cmssw/search?utf8=%E2%9C%93&q=GeometryExtended2023LRecoReco_cff

still refer to GeometryExtended2023LRecoReco_cff. Is there any PR which fixes these?

**Comment by ianna** [comment]
@smuzaffar - this is it:
https://github.com/cms-sw/cmssw/pull/14906
